streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareTaskAssignorTest.java

```diff
@@ -715,23 +1087,49 @@ private Map<UUID, Map<String, Optional<String>>> getRandomProcessRacks(final int
         return processRacks;
     }
 
-    private SortedMap<TaskId, Set<TopicPartition>> getTaskTopicPartitionMap(final int tpSize) {
+    private SortedMap<TaskId, Set<TopicPartition>> getTaskTopicPartitionMap(final int tpSize, final boolean changelog) {
         final SortedMap<TaskId, Set<TopicPartition>> taskTopicPartitionMap = new TreeMap<>();
+        final String topicName = changelog ? CHANGELOG_TOPIC_PREFIX : TOPIC_PREFIX;
         for (int i = 0; i < tpSize; i++) {
-            taskTopicPartitionMap.put(new TaskId(i, 0), mkSet(new TopicPartition("topic" + i, 0)));
+            taskTopicPartitionMap.put(new TaskId(i, 0), mkSet(
+                new TopicPartition(topicName + i, 0),
+                new TopicPartition(topicName + (i + 1) % tpSize, 0)
+            ));
         }
         return taskTopicPartitionMap;
     }
 
-    private SortedMap<UUID, ClientState> getRandomClientState(final int clientSize, final int tpSize) {
+    private InternalTopicManager mockInternalTopicManagerForRandomChangelog(final int nodeSize, final int tpSize) {
+        final Set<String> changelogNames = new HashSet<>();
+        final List<Node> nodeList = getRandomNodes(nodeSize);
+        final Map<String, List<TopicPartitionInfo>> topicPartitionInfo = new HashMap<>();
+        for (int i = 0; i < tpSize; i++) {
+            final String topicName = CHANGELOG_TOPIC_PREFIX + i;
+            changelogNames.add(topicName);
+
+            final Node firstNode = nodeList.get(i % nodeSize);
+            final Node secondNode = nodeList.get((i + 1) % nodeSize);
+            final TopicPartitionInfo info = new TopicPartitionInfo(0, firstNode, Arrays.asList(firstNode, secondNode), Collections.emptyList());
+
+            topicPartitionInfo.computeIfAbsent(topicName, tp -> new ArrayList<>()).add(info);
+        }
+
+        final MockInternalTopicManager spyTopicManager = spy(mockInternalTopicManager);
+        doReturn(topicPartitionInfo).when(spyTopicManager).getTopicPartitionInfo(changelogNames);
+        return spyTopicManager;
+    }
+
+    private SortedMap<UUID, ClientState> getRandomClientState(final int clientSize, final int tpSize, final int maxCapacity) {
         final SortedMap<UUID, ClientState> clientStates = new TreeMap<>();
         final List<TaskId> taskIds = new ArrayList<>(tpSize);
         for (int i = 0; i < tpSize; i++) {
             taskIds.add(new TaskId(i, 0));
         }
         Collections.shuffle(taskIds);
+        final Random random = new Random();
```

Comment:
We should generate our own seed and print the seed -- in case a test fails, we can use the seed to reproduce the issue.